### ISSUE TITLE
Error when getting a false value

### ISSUE THREAD

**Issue by elvogel** [microsoft/RulesEngine#40]
I have the following workflow rule(s) defined:
```json
[
  {
    "WorkflowName": "ConsVendCertificateEntity",
    "Rules": [
      {
        "RuleName": "VendorPartyNumberCheck",
        "SuccessEvent": "10",
        "ErrorMessage": "VendorPartyNumber doesn't exist in Dynamics",
        "ErrorType": "Error",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "Utils.ValueExists(\"ConsVendCertificateEntity\",\"VendorPartyNumber\",input1.VENDORPARTYNUMBER) == true"
      },
      {
        "RuleName": "VendorCertificateType",
        "SuccessEvent": "15",
        "WarningMessage": "VendorCertificateType isn't General Liability or Workmen's Compensation",
        "ErrorType": "Warning",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "Utils.CheckContains(input1.VENDCERTIFICATETYPENAME,\"General Liability,Workmen's Compensation\") == true"
      }
    ]
  }
]
```
I loop through records in an Azure Storage table to validate data in them. The `Utils.CheckContains` method is from the Wiki documentation. My code (abbreviated) looks like this:

```csharp
                foreach (var rule in workflowRules)
                {
                    List<RuleResultTree> results;
                    try
                    {
                        results = engine.ExecuteRule(rule.WorkflowName, input1);
                    }
                    catch (Exception e)
                    {
                        log.LogError($"[{tableName}] {e.GetType()}: {e.Message}\r\n{e.StackTrace}");
                        throw new ApplicationException($"An exception was thrown in the rules engine at row {row} for {tableName}. Check the rules and try again.");
                    }
               }
```
When I get through the first record, it passes both rules. When I get to the second record, it fails the `VendorCertificateType` rule because my record has `Workmen's Compensation` spelled wrong. It's not null. But instead of getting a rule failure, I get this exception:

```csharp
ArgumentNullException: Value cannot be null. (Parameter 'input')
   at System.Text.RegularExpressions.Regex.Matches(String input)
   at System.Text.RegularExpressions.Regex.Matches(String input, String pattern, RegexOptions options, TimeSpan matchTimeout)
   at System.Text.RegularExpressions.Regex.Matches(String input, String pattern)
   at RulesEngine.RulesEngine.FormatErrorMessages(IEnumerable`1 result)
   at RulesEngine.RulesEngine.ExecuteRuleByWorkflow(String workflowName, RuleParameter[] ruleParameters)
   at RulesEngine.RulesEngine.ValidateWorkflowAndExecuteRule(String workflowName, RuleParameter[] ruleParams)
   at RulesEngine.RulesEngine.ExecuteRule(String workflowName, RuleParameter[] ruleParams)
   at RulesEngine.RulesEngine.ExecuteRule(String workflowName, Object[] inputs)
   at SM.Integration.Repos.BRE.RulesProcessor.ProcessRules() in C:\path\to\my\code\RulesProcessor.cs:line 88
```
The `Utils.CheckContains` code, for reference

```csharp
        public static bool CheckContains(string check, string valList)
        {
            if (string.IsNullOrEmpty(check) || string.IsNullOrEmpty(valList))
                return false;

            var list = valList.Split(',').ToList();
            return list.Contains(check);
        }
```

**Comment by abbasc52** [comment]
@elvogel  It seems this bug was introduced in v2.1.0, we will fix it as soon as possible. Could you try and see if you can reproduce it in v2.0.0? 

Also if possible please provide a sample input so we can reproduce this error.

**Comment by elvogel** [comment]
I rolled back to v2.0.0. The rule above works; however, a new rule now starts complaining. 

When I run this rule:
```json
[
  {
    "WorkflowName": "ConsItemPriceSalesEntity",
    "Rules": [
      {
        "RuleName": "ItemIDCheck",
        "SuccessEvent": "10",
        "ErrorMessage": "ItemID is not Option or LotPremium",
        "ErrorType": "Error",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "Utils.CheckContains(input1.ITEMID,\"Option,LotPremium\") == true"
      },
      {
        "RuleName": "ElementNameExists",
        "SuccessEvent": "15",
        "ErrorMessage": "ElementName is null",
        "ErrorType": "Error",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "item1.ELEMENTNAME != null"
      }
    ]
  }
]
```

I get this error:
```csharp
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.

   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at RulesEngine.RulesEngine.ExecuteRuleByWorkflow(String workflowName, RuleParameter[] ruleParams)
   at RulesEngine.RulesEngine.ValidateWorkflowAndExecuteRule(String workflowName, RuleParameter[] ruleParams)
   at RulesEngine.RulesEngine.ExecuteRule(String workflowName, RuleParameter[] ruleParams)
   at RulesEngine.RulesEngine.ExecuteRule(String workflowName, Object[] inputs)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite site, T0 arg0, T1 arg1, T2 arg2)
   at SM.Integration.Repos.BRE.RulesProcessor.ProcessRules() in C:\path\to\my\code\RulesProcessor.cs:line 88
```

The inner exception might be helpful here as well:

```csharp
System.InvalidOperationException: Target object is not an ExpandoObject
   at System.Linq.Dynamic.Core.DynamicGetMemberBinder.FallbackGetMember(DynamicMetaObject target, DynamicMetaObject errorSuggestion)
   at System.Dynamic.GetMemberBinder.FallbackGetMember(DynamicMetaObject target)
   at System.Dynamic.DynamicMetaObject.BindGetMember(GetMemberBinder binder)
   at System.Dynamic.GetMemberBinder.Bind(DynamicMetaObject target, DynamicMetaObject[] args)
   at System.Dynamic.DynamicMetaObjectBinder.Bind(Object[] args, ReadOnlyCollection`1 parameters, LabelTarget returnLabel)
   at System.Runtime.CompilerServices.CallSiteBinder.BindCore[T](CallSite`1 site, Object[] args)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
   at lambda_method(Closure , <>f__AnonymousType0`19 , RuleInput )
```
This time, I'm getting a value of **true** from `Utils.CheckContains`.  When I step into the utility method, the values exist, and it operates as expected.

**Comment by abbasc52** [comment]
@elvogel  i see you are using `"Expression": "item1.ELEMENTNAME != null"` are you passing a RuleParameter with `item1` alias?
This might be related to #30 .

**Comment by elvogel** [comment]
Yes I am. I have an ExpandoObject called item1 that gets assigned. All the rules reuse the same code base.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by abbasc52** [comment]
Thanks for the reply @elvogel . Can you share some sample input so that we can reproduce this issue?

**Comment by elvogel** [comment]
I can't give you customer data - however, I can give you the classes that I use for converting it into an `ExpandoObject` below:

DynamicTableEntityJsonConverter:
```csharp
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using Microsoft.Azure.Cosmos.Table;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace SM.Dynamics.Common.Serialization
{
public class DynamicTableEntityJsonConverter : JsonConverter
  {
    private readonly List<string> excludedProperties;

    public DynamicTableEntityJsonConverter(List<string> excludedProperties = null)
    {
      this.excludedProperties = excludedProperties;
    }

    public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
    {
      if (value == null)
        return;
      writer.WriteStartObject();
      WriteJsonProperties(writer, (DynamicTableEntity) value, excludedProperties);
      writer.WriteEndObject();
    }

    public override object ReadJson(
      JsonReader reader,
      Type objectType,
      object existingValue,
      JsonSerializer serializer)
    {
      if (reader.TokenType == JsonToken.Null)
        return null;
      DynamicTableEntity dynamicTableEntity = new DynamicTableEntity();

      using List<JProperty>.Enumerator enumerator = JObject.Load(reader).Properties().ToList().GetEnumerator();
      while (enumerator.MoveNext())
      {
        JProperty current = enumerator.Current;
        if (current == null) continue;
        if (string.Equals(current.Name, "PartitionKey", StringComparison.Ordinal))
          dynamicTableEntity.PartitionKey = current.Value.ToString();
        else if (string.Equals(current.Name, "RowKey", StringComparison.Ordinal))
          dynamicTableEntity.RowKey = current.Value.ToString();
        else if (string.Equals(current.Name, "Timestamp", StringComparison.Ordinal))
          dynamicTableEntity.Timestamp = (DateTimeOffset) current.Value;
        else if (string.Equals(current.Name, "ETag", StringComparison.Ordinal))
        {
          dynamicTableEntity.ETag = current.Value.ToString();
        }
        else
        {
          var entityProperty = CreateEntityProperty(serializer, current);
          dynamicTableEntity.Properties.Add(current.Name, entityProperty);
        }
      }

      return dynamicTableEntity;
    }

    public override bool CanConvert(Type objectType) => typeof (DynamicTableEntity).IsAssignableFrom(objectType);

    private static void WriteJsonProperties(
      JsonWriter writer,
      DynamicTableEntity entity,
      List<string> excludedProperties = null)
    {
      if (entity == null)
        return;
      writer.WritePropertyName("PartitionKey");
      writer.WriteValue(entity.PartitionKey);
      writer.WritePropertyName("RowKey");
      writer.WriteValue(entity.RowKey);
      writer.WritePropertyName("Timestamp");
      writer.WriteValue(entity.Timestamp);
      writer.WritePropertyName("ETag");
      writer.WriteValue(entity.ETag);

      using IEnumerator<KeyValuePair<string, EntityProperty>>
        enumerator = (excludedProperties == null ? entity.Properties :
          entity.Properties.Where(p => !excludedProperties.Contains(p.Key))).GetEnumerator();
      while (enumerator.MoveNext())
      {
        KeyValuePair<string, EntityProperty> current = enumerator.Current;
        WriteJsonProperty(writer, current);
      }
    }

    private static void WriteJsonProperty(
      JsonWriter writer,
      KeyValuePair<string, EntityProperty> property)
    {
      if (string.IsNullOrWhiteSpace(property.Key) || property.Value == null)
        return;
      switch ((int) property.Value.PropertyType)
      {
        case 0:
          WriteJsonPropertyWithEdmType(writer, property.Key, property.Value.StringValue, property.Value.PropertyType);
          break;
        case 1:
          WriteJsonPropertyWithEdmType(writer, property.Key, property.Value.BinaryValue, property.Value.PropertyType);
          break;
        case 2:
          WriteJsonPropertyWithEdmType(writer, property.Key, property.Value.BooleanValue, property.Value.PropertyType);
          break;
        case 3:
          WriteJsonPropertyWithEdmType(writer, property.Key, property.Value.DateTimeOffsetValue, property.Value.PropertyType);
          break;
        case 4:
          WriteJsonPropertyWithEdmType(writer, property.Key, property.Value.DoubleValue, property.Value.PropertyType);
          break;
        case 5:
          WriteJsonPropertyWithEdmType(writer, property.Key, property.Value.GuidValue, property.Value.PropertyType);
          break;
        case 6:
          WriteJsonPropertyWithEdmType(writer, property.Key, property.Value.Int32Value, property.Value.PropertyType);
          break;
        case 7:
          WriteJsonPropertyWithEdmType(writer, property.Key, property.Value.Int64Value, property.Value.PropertyType);
          break;
        default:
          throw new NotSupportedException(string.Format(CultureInfo.InvariantCulture, "Unsupported EntityProperty.PropertyType:{0} detected during serialization.", property.Value.PropertyType));
      }
    }

    private static void WriteJsonPropertyWithEdmType(
      JsonWriter writer,
      string key,
      object value,
      EdmType edmType)
    {
      //writer.WritePropertyName(key);
      //writer.WriteStartObject();
      writer.WritePropertyName(key);
      writer.WriteValue(value);
      // writer.WritePropertyName("EdmType");
      // writer.WriteValue(edmType.ToString());
      //writer.WriteEndObject();
    }

    private static EntityProperty CreateEntityProperty(
      JsonSerializer serializer,
      JProperty property)
    {
      if (property == null)
        return null;
      List<JProperty> list = JObject.Parse(property.Value.ToString()).Properties().ToList();
      EdmType edmType = (EdmType) Enum.Parse(typeof (EdmType), list[1].Value.ToString()!, true);
      EntityProperty entityProperty;
      switch ((int) edmType)
      {
        case 0:
          entityProperty = EntityProperty.GeneratePropertyForString(list[0].Value.ToObject<string>(serializer));
          break;
        case 1:
          entityProperty = EntityProperty.GeneratePropertyForByteArray(list[0].Value.ToObject<byte[]>(serializer));
          break;
        case 2:
          entityProperty = EntityProperty.GeneratePropertyForBool(list[0].Value.ToObject<bool>(serializer));
          break;
        case 3:
          entityProperty = EntityProperty.GeneratePropertyForDateTimeOffset(list[0].Value.ToObject<DateTimeOffset>(serializer));
          break;
        case 4:
          entityProperty = EntityProperty.GeneratePropertyForDouble(list[0].Value.ToObject<double>(serializer));
          break;
        case 5:
          entityProperty = EntityProperty.GeneratePropertyForGuid(list[0].Value.ToObject<Guid>(serializer));
          break;
        case 6:
          entityProperty = EntityProperty.GeneratePropertyForInt(list[0].Value.ToObject<int>(serializer));
          break;
        case 7:
          entityProperty = EntityProperty.GeneratePropertyForLong(list[0].Value.ToObject<long>(serializer));
          break;
        default:
          throw new NotSupportedException(string.Format(CultureInfo.InvariantCulture,
            "Unsupported EntityProperty.PropertyType:{0} detected during deserialization.", edmType));
      }
      return entityProperty;
    }
  }
}

DynamicTableEntityJsonSerializer:

```csharp
using System.Collections.Generic;
using Microsoft.Azure.Cosmos.Table;
using Newtonsoft.Json;

namespace SM.Dynamics.Common.Serialization
{
    public class DynamicTableEntityJsonSerializer
    {
        private readonly DynamicTableEntityJsonConverter jsonConverter;

        public DynamicTableEntityJsonSerializer(List<string> excludedProperties = null) => this.jsonConverter = new DynamicTableEntityJsonConverter(excludedProperties);

        public string Serialize(DynamicTableEntity entity)
        {
            return entity != null ? JsonConvert.SerializeObject(entity, jsonConverter) : null;
        }

        public DynamicTableEntity Deserialize(string serializedEntity)
        {
            return serializedEntity != null ?
                JsonConvert.DeserializeObject<DynamicTableEntity>(serializedEntity, (JsonConverter) jsonConverter) :
                null;
        }

    }
}

```
RulesProcessor.cs
```csharp
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Azure.Cosmos.Table;
using Microsoft.Azure.Storage.Blob;
using Microsoft.Azure.WebJobs;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using RulesEngine.Models;
using SM.Dynamics.Common.Serialization;

namespace SM.Integration.Repos.BRE
{
    public class RulesProcessor
    {
        private readonly ILogger log;
        private readonly CloudTable table;
        protected readonly CloudBlobContainer ruleContainer;
        protected readonly List<WorkflowRules> workflowRules;
        protected readonly CloudBlobClient client;
        protected readonly global::RulesEngine.RulesEngine engine;
        protected readonly CloudTableClient tableClient;
        protected readonly string tableName;
        protected readonly StorageAccount account =
            StorageAccount.NewFromConnectionString(
                Environment.GetEnvironmentVariable(Constants.StorageConnection));

        public RulesProcessor(ILogger logger, string tableName)
        {
            log = logger;
            tableClient = account.CreateCloudTableClient();
            client = account.CreateCloudBlobClient();

            table = tableClient.GetTableReference(tableName);
            ruleContainer = client.GetContainerReference("rules");

            this.tableName = tableName;

            var blobRef = ruleContainer.GetBlobReference($"{tableName}.json");

            if (!blobRef.Exists())
            {
                var s = $"Rules do not exist for entity {tableName} in rules container.";
                log.LogWarning(s);
                throw new ApplicationException(s);
            }
            var sr = new StreamReader(blobRef.OpenRead());
            var ruleText = sr.ReadToEnd();
            workflowRules = JsonConvert.DeserializeObject<List<WorkflowRules>>(ruleText);
            engine = new global::RulesEngine.RulesEngine(workflowRules.ToArray(), log,
                new ReSettings{CustomTypes = new []{typeof(Utils)}});
        }

        public Task ProcessRules()
        {
            TableContinuationToken token = null;
            var entities = new List<DynamicTableEntity>();
            do
            {
                var qr = table.ExecuteQuerySegmented(new TableQuery<DynamicTableEntity>(), token);
                entities.AddRange(qr.Results);
                token = qr.ContinuationToken;
            } while (token != null);

            var row = 1;
            foreach (var entity in entities)
            {
                var serializer = new DynamicTableEntityJsonSerializer();
                var converter = new ExpandoObjectConverter();

                //serialize the entity, then deserialize as an ExpandoObject
                var data = serializer.Serialize(entity);
                dynamic input1 = JsonConvert.DeserializeObject<ExpandoObject>(data, converter);

                // var ruleParams =
                //     entity.Properties.Select(property =>
                //         new RuleParameter(property.Key, property.Value.PropertyAsObject)).ToList();

                foreach (var rule in workflowRules)
                {
                    List<RuleResultTree> results;
                    try
                    {
                        results = engine.ExecuteRule(rule.WorkflowName, input1);
                    }
                    catch (Exception e)
                    {
                        log.LogError($"[{tableName}] {e.GetType()}: {e.Message}\r\n{e.StackTrace}");
                        throw new ApplicationException($"An exception was thrown in the rules engine at row {row} for {tableName}. Check the rules and try again.");
                    }

                    foreach (var result in results)
                    {
                        if (result.IsSuccess) continue;

                        var sb = new StringBuilder();
                        sb.AppendLine($"{rule.WorkflowName}.{result.Rule} encountered an error in row {row}: {result.ExceptionMessage}");
                        var msg = result.GetMessages();
                        foreach (var warn in msg.WarningMessages)
                        {
                            sb.AppendLine($"Warning: {warn}");
                        }

                        foreach (var errorMessage in msg.ErrorMessages)
                        {
                            sb.AppendLine($"Error: {errorMessage}");
                        }
                        log.LogError(sb.ToString());
                        log.LogError($"processing for {tableName} will now stop due to errors.");
                        throw new ApplicationException($"Processing for {tableName} stopped because rule {result.Rule.RuleName} failed to pass.");
                    }
                }

                row++;
            }
            log.LogInformation($"Rules processed successfully for {tableName}");
            return Task.CompletedTask;
        }
    }
}
```

**Comment by abbasc52** [comment]
@elvogel Looking at your implementation, it seems you use the default naming provided by ruleengine(input1,input2,etc)

In your workflow which gives error i  can see two names used for input object `input1` in rule `ItemIDCheck` and `item1` for rule `ElementNameExists`

```
[
  {
    "WorkflowName": "ConsItemPriceSalesEntity",
    "Rules": [
      {
        "RuleName": "ItemIDCheck",
        "SuccessEvent": "10",
        "ErrorMessage": "ItemID is not Option or LotPremium",
        "ErrorType": "Error",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "Utils.CheckContains(input1.ITEMID,\"Option,LotPremium\") == true"
      },
      {
        "RuleName": "ElementNameExists",
        "SuccessEvent": "15",
        "ErrorMessage": "ElementName is null",
        "ErrorType": "Error",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "item1.ELEMENTNAME != null" 
      }
    ]
  }
]
```

The exception you are getting is due to ruleEngine unable to find `item1` in the passed input. 
If you intend to pass 2 input objects with specific naming you can do it as follows:
```csharp
results = engine.ExecuteRule(rule.WorkflowName, new RuleParameter("input1",input1), new RuleParameter("item1", item1));
```

I agree we need to provide a better error message in this scenario and we are currently working on fixing it.

On a side note: RuleEngine supports typed classes as input as well 😄

**Comment by elvogel** [comment]
Whoops! Okay, skip that 'second error'. Thank you for the find!

**Comment by abbasc52** [comment]
This issue is fixed in v2.1.1
